### DISCUSSION TITLE
BME280 temperature sensor's code

### DISCUSSION THREAD

**Discussion by fhwe71** [arendst/Tasmota#11462] [Q&A] [answered]
Hi all,
I am quite new to Tasmota and am still in evaluation phase. Thus I bought a Sonoff Dual R2 to play with, compiled Tasmota for a BME280 and connected it via I2C. It works, but values are not quite precise. This is an issue I read quite often in internet forums. Therefore I compared it with the official Bosch code (https://github.com/BoschSensortec/BME280_driver) running on a Raspberry. 

The result is amazing: The Bosch code delivers very precise values, Tasmota does not.
I guess either the measurement algorithm by Adafruit, the calibration or device mode handling have a bug. 
Any experiences about that? 

That brings me to my second question: How to develop or debug Tasmota?
Changing the code in vscode, compiling, flashing the Sonoff, adding device to Wifi, rewire the sensor, configure it... Every round takes 10-15 minutes and I guess a code optimization takes many rounds...
Better do this on NodeMCU or similar, or isolate the code to a Raspberry Pi?

Kind regards,
 Felix

**Comment by Jason2866** [comment, marked as answer]
Can not reproduce your experience. I have several BME280 and all are very close with there values.
Temperature measured with a calibrated thermometer.
Btw. Tasmota does not use Adafruit library. The code is based on Adafruit and modified.

**Reply by fhwe71** [reply]
Ok, I will give it a second try. My finding was Bosch code showed same results as my thermometer of trust, the modified Adafruit code 2-3 degree Celsius below that value. Same for humidity. Pressure I cannot measure for myself, but both Bosch and Adafruit code do not fit to wheather forecast. Always using the very same BME280 sensor.

**Reply by Milbert-R** [reply]
The indication of air pressure in weather forecasts is usually the air pressure at sea level for better comparability.

With the command "altitude", Tasmota offers the option of entering the altitude of your own location via sea level. With a BME280, after entering a value for "altitude", you will also receive information on the sea pressure.
 
In order to compare the information with the weather forecast, you need to know where the weather station for the weather forecast is located and enter the corresponding value for "altitude".
![deepsleep_test](https://user-images.githubusercontent.com/37844838/112519463-af45cf80-8d9a-11eb-9a78-ba2cefaf7589.png)

**Reply by sfromis** [reply]
As the Tasmota driver for BME280 does contain more than a little bit of [calculation and calibration](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_09_bmp.ino#L294-L343), I suppose that someone could verify how "good" that code is, compared to the vendor code.

**Reply by fhwe71** [reply]
Thanks, did not know this. Will set it accordingly.

**Reply by fhwe71** [reply]
> As the Tasmota driver for BME280 does contain more than a little bit of [calculation and calibration](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_09_bmp.ino#L294-L343), I suppose that someone could verify how "good" that code is, compared to the vendor code.

I have seen that. Will try to isolate the measurement/calibration part and test it systematically compared to vendor code. Might take some days.

**Comment by barbudor** [comment]
> That brings me to my second question: How to develop or debug Tasmota?
> Changing the code in vscode, compiling, flashing the Sonoff, adding device to Wifi, rewire the sensor, configure it... Every round takes > 10-15 minutes and I guess a code optimization takes many rounds...
> Better do this on NodeMCU or similar, or isolate the code to a Raspberry Pi?

Why dismantling ? Once you've compiled just flash OTA and that's it.
I don't care where is my device, it could be in another room or on the other side of the planet. As far as I can send the OtaUpgrade command I don't need to be close to it.
Now developping on a NodeMCU or mini-D1 is interresting as many reflash will wear off the flash faster. So better use a disposable board that you can throw to the bin at the 1st doubt.
Still working on the final device may be required if you can't on a node MCU such as working with AC voltage, accessing specific hardware....

**Reply by fhwe71** [reply]
Ok, somehow I was blind. I considered OTA as a way for pre-produced binaries, not for stuff compiled by me. Definitely easier like that, thanks. 
I did not consider wear-off yet. Any experience about that? During development, I flash my ATmega many times a day without any errors so far. Are ESP chips more critical regarding wear-off?

**Reply by sfromis** [reply]
It is not the ESP chips, but the flash chips, which are basically the same as for other MCU platforms. Normally it is not too much of an issue unless updating flash very, very often. Still developing on a disposable board is generally a good plan, stuff can happen, also outside of flash considerations.

**Reply by barbudor** [reply]
But sonoff devices like many commercial ones use ESP8285 with flash inside the MCU chip. Not changeable in case of problem.
That's why I would advise to use a s much as possible Nodemcu which, as an Arduino board, is disposable.
Flash issue? Trash
Contact problem? Trash
They cost like 20€ per 10 units on Ali and ESP32 de kit as not much more expensive
